### PR TITLE
Closes #2992 : Add per-doc revs_limit to the _bulk_get endpoint

### LINKED ISSUE #2992: Support per-doc revs_limit for bulk_get
Currently Sync Gateway only looks for the revs_limit parameter as a top-level query parameter.  

Couchbase Lite is sending this parameter per document (in the `docs` array).  Sync Gateway should support either the top level or the docs level property.

We should look for a revs_limit parameter in the document here:
https://github.com/couchbase/sync_gateway/blob/master/rest/bulk_api.go#L397

and if present, use that instead of the existing revsLimit value.

Sample query that sets revs_limit:
```
POST /<bucketname>/_bulk_get?revs=true&attachments=true HTTP/1.1 (application/json)
{"docs":[{"revs_limit":5,"rev":"51-
a3881d1fd4540de8027d9141322d2019","id":"c8b1d19a573c665a08124ee1475fcd57"},{"r
evs_limit":5,"rev":"51-
02663483dbac053d215316f8af7c3dc2","id":"328b91bbc0b7b49848bb18c21f8ff0da"},{"rev
s_limit":5,"rev":"51-
b93e8ae6fb02f74dbb79d0022fb0c5c2","id":"eaa38672490a50198c27e4b322297d38]}
```

### PR DESCRIPTION
**Pull request by bbrks**
Closes #2992 : Add per-doc revs_limit to the _bulk_get endpoint.

# Integration Tests
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/262/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/263/

### CODE REVIEW

File: rest/bulk_api.go [hunk 1]
```diff
@@ -397,21 +393,35 @@ func (h *handler) handleBulkGet() error {
...
-				if revsLimit > 0 {
-					revsFrom, err = db.GetStringArrayProperty(doc, "revs_from")
-					if revsFrom == nil {
-						revsFrom = attsSince // revs_from defaults to same value as atts_since
+
+				if h.getBoolQuery("revs") {
+					// Try to pull out a per-doc revs limit, otherwise use request limit.
+					if raw, ok := doc["revs_limit"]; !ok {
+						docRevsLimit = int(h.getIntQuery("revs_limit", math.MaxInt32))
+					} else if val, ok := raw.(float64); ok && val >= 0 {
```
Comment (adamcfraser) [inline rest/bulk_api.go]: I think using base.ToInt64 would be preferable here.

File: rest/bulk_api.go [hunk 2]
```diff
@@ -397,21 +393,35 @@ func (h *handler) handleBulkGet() error {
...
+						docRevsLimit = int(val)
+					} else {
+						err = base.HTTPErrorf(http.StatusBadRequest, "Invalid revs_limit for doc: %s in _bulk_get", docid)
```
Comment (adamcfraser) [inline rest/bulk_api.go]: I found the logic here a bit hard to parse as if/else if/else, with multiple usages of ok.  I think it would be clearer with the second else nested, to make it clearer that it's associated with the case where revs_limit is present (but invalid).

File: rest/bulk_api.go [hunk 3]
```diff
@@ -397,21 +393,35 @@ func (h *handler) handleBulkGet() error {
 				err = base.HTTPErrorf(http.StatusBadRequest, "Invalid doc/rev ID in _bulk_get")
 			} else {
 				attsSince, err = db.GetStringArrayProperty(doc, "atts_since")
-				if revsLimit > 0 {
-					revsFrom, err = db.GetStringArrayProperty(doc, "revs_from")
-					if revsFrom == nil {
-						revsFrom = attsSince // revs_from defaults to same value as atts_since
+
+				if h.getBoolQuery("revs") {
```
Comment (adamcfraser) [inline rest/bulk_api.go]: It would be better to read the query level properties ("revs", "revs_limit") once prior to ranging over docs, instead of re-parsing from the query string in each iteration.

File: rest/api_test.go [hunk 4]
```diff
@@ -1348,6 +1348,126 @@ func TestOpenRevs(t *testing.T) {
 ]`)
 }
 
+// Attempts to get a varying number of revisions on a per-doc basis.
+// Covers feature implemented in issue #2992
+func TestBulkGetPerDocRevsLimit(t *testing.T) {
```
Comment (adamcfraser) [inline rest/api_test.go:1353]: Nice test!  Very thorough.

**Review by adamcfraser — CHANGES_REQUESTED** [review]
Looks good functionally - a couple of smaller style/performance comments.

File: rest/api_test.go [hunk 5]
```diff
@@ -1235,11 +1235,11 @@ func TestBulkDocsChangeToRoleAccess(t *testing.T) {
...
 				{
-					"_id": "bulk1", 
-				 	"type" : "roleaccess", 
-				 	"grantTo":"role:role1", 
+					"_id": "bulk1",
+				 	"type" : "roleaccess",
+				 	"grantTo":"role:role1",
 				 	"grantChannel":"chan1"
-				 }, 
+				 },
```
[the remark this answers is not in the document]
Comment (bbrks) [reply rest/api_test.go:1242]: I thought it was gofmt, but it looks like it's down to my editor settings. I'll turn it off for now.

I'm running [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports) on save, but it will use the same formatting rules as gofmt. After reading into it a little more, gofmt will preserve any leading/trailing whitespace.

File: rest/bulk_api.go (hunk 2 above)
Comment (bbrks) [reply rest/bulk_api.go]: Agreed. Thanks for the review, I'll update today.

**Review by adamcfraser — APPROVED** [review]
